dotCMS/src/main/java/com/dotcms/api/content/VanityUrlFactoryImpl.java

```diff
@@ -0,0 +1,226 @@
+		List<VanityUrl> results = new ArrayList<VanityUrl>();
+		List<Contentlet> contentResults = contentletAPI.search("+baseType:"+BaseContentType.VANITY_URL.getType()+" +live:true +deleted:false", 0, 0, "", user, false);
+		contentResults.stream().forEach((Contentlet con) -> {
+			VanityUrl vanityUrl = fromContentlet(con);
+			addToVanityURLCache(vanityUrl);
+			results.add(vanityUrl);
+		});
+		
+		return results;
+	}
+
+	@Override
+	public VanityUrl getVanityUrlByURI(String uri, Host host, long languageId, User user, boolean live) 
+			throws DotDataException, DotSecurityException {
+		VanityUrl result = vanityURLCache.get(VanityUrlUtil.sanitizeKey(host != null && InodeUtils.isSet(host.getInode())?host.getHostname()+"|"+uri:uri,languageId));
+		if(result == null || !InodeUtils.isSet(result.getInode())){
+			List<VanityUrl> results = new ArrayList<VanityUrl>();
+			String hostCondition = (host != null?host.getHostname():"");
+			List<Contentlet> contentResults = contentletAPI.search("+baseType:"+BaseContentType.VANITY_URL.getType()+" +languageId:"+languageId+" +vanityUrl:"+hostCondition+uri+(live?" +live:true":" +working:true"), 0, 0, "", user, false);
+			contentResults.stream().forEach((Contentlet con) ->{
+						VanityUrl vanityUrl = fromContentlet(con);
+						addToVanityURLCache(vanityUrl);
+						results.add(vanityUrl);
+					});
+
+			if(results.size() == 0 && Config.getBooleanProperty("DEFAULT_VANITY_URL_TO_DEFAULT_LANGUAGE", false)){
+				contentResults = contentletAPI.search("+baseType:"+BaseContentType.VANITY_URL.getType()+" +languageId:"+APILocator.getLanguageAPI().getDefaultLanguage().getId()+" +vanityUrl:"+hostCondition+uri+(live?" +live:true":" +working:true"), 0, 0, "", user, false);
+				contentResults.stream().forEach((Contentlet con) ->{
+							VanityUrl vanityUrl = fromContentlet(con);
+							addToVanityURLCache(vanityUrl);
+							results.add(vanityUrl);
+						});
+			}
+			result = results.size() > 0?results.get(0):null;
+		}
+
+		return result;
+	}
+
+	/**
+	 * Add the the search results to the vanityURLCache
+	 * @param results
+	 */
+	private void addToVanityURLCache(VanityUrl vanityUrl){
+		try {
+			vanityURLCache.add(VanityUrlUtil.sanitizeKey(vanityUrl), vanityUrl);
+		} catch (DotDataException | DotRuntimeException | DotSecurityException e) {
+			Logger.error(this, "Error trying to add Vanity URL identifier:"+vanityUrl.getIdentifier()+" to VanityURLCache",e);
+		}
+	}
+
+	/**
+	 * Add the the search results to the vanityURLCache
+	 * @param results
+	 */
+	private void removeFromVanityURLCache(VanityUrl vanityUrl){
+		try {
+			vanityURLCache.remove(VanityUrlUtil.sanitizeKey(vanityUrl));
+		} catch (DotDataException | DotRuntimeException | DotSecurityException e) {
+			Logger.error(this, "Error trying to add Vanity URL identifier:"+vanityUrl.getIdentifier()+" to VanityURLCache",e);
+		}
+	}
+
+	@Override
+	public VanityUrl fromContentlet(Contentlet con) {
+
+		if (con != null){
```

Comment:
@oswaldogallango fix me if I am wrong but I think you are calling the fromContentlet as a fallback in the cache, I mean the cache is who call the fromContentlet if the Vanity is not in the cache.

Also, I think all Vanity's are in cache preloaded at the begin of the App.

Also the method could be named, getVanityUrlFromContentLet :) i